File: ui_c/src/os/os_gfx.c
#include "os_gfx.h"

#include <assert.h>

void
os_gfx_init() {
  assert(glfwInit());
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
}

Window
os_window_open(Vec2F32 resolution, String8 title) {
  glfwWindowHint(GLFW_DECORATED, GLFW_FALSE);

  GLFWwindow* window = glfwCreateWindow(resolution.x, resolution.y,
                                        (const char*)title.str, NULL, NULL);
  if (!window) {
    glfwTerminate();
    assert(window && "Could not create window");
  }
  glfwMakeContextCurrent(window);
  glfwSwapInterval(0);

  glewExperimental = GL_TRUE;
  assert(glewInit() == GLEW_OK && "Could not create glew bindings");

  return (Window){.handle = window};
}

bool
os_window_should_close(Window* window) {
  bool should_close =
      (glfwGetKey(window->handle, GLFW_KEY_ESCAPE) == GLFW_PRESS);
  if (should_close) {
    glfwSetWindowShouldClose(window->handle, 1);
  }
  return should_close;
}

File: zig-eval/src/os/os_gfx.zig
const std = @import("std");
const base = @import("../base/base_core.zig");
const assert = std.debug.assert;
const c = @cImport({
    @cInclude("glad.h");
    @cInclude("GLFW/glfw3.h");
});

const OS_Window = struct {
    handle: *c.GLFWwindow,
    size: base.Vec2(i32),
    pos: base.Vec2(i32),
};

/// initialize os layer, call once.
pub fn gfx_init() void {
    assert(c.glfwInit() == c.GLFW_TRUE);
    c.glfwWindowHint(c.GLFW_CONTEXT_VERSION_MAJOR, 3);
    c.glfwWindowHint(c.GLFW_CONTEXT_VERSION_MINOR, 3);
    c.glfwWindowHint(c.GLFW_OPENGL_PROFILE, c.GLFW_OPENGL_CORE_PROFILE);
}

pub fn window_open(resolution: base.Vec2(i32)) OS_Window {
    c.glfwWindowHint(c.GLFW_DECORATED, c.GL_FALSE);
    const window =
        c.glfwCreateWindow(resolution.x, resolution.y, "", null, null);
    if (window) |w| {
        c.glfwMakeContextCurrent(w);
        c.glfwSwapInterval(0);

        assert(c.gladLoadGLLoader(@as(c.GLADloadproc, @ptrCast(&c.glfwGetProcAddress))) != 0);

        var os_window = std.mem.zeroInit(OS_Window, .{ .handle = w });
        c.glfwGetWindowSize(os_window.handle, &os_window.size.x, &os_window.size.y);
        c.glfwGetWindowPos(os_window.handle, &os_window.pos.x, &os_window.pos.y);
        return os_window;
    } else {
        c.glfwTerminate();
        @panic("Could not create window");
    }
}
